frb_rust/src/ffi/io.rs

```diff
@@ -30,3 +33,154 @@ mod tests {
+/// be safely sent to Dart. For example, this declaration can be used across the
+/// FFI border:
+/// ```rust
+/// use flutter_rust_bridge::*;
+/// use std::fmt::Debug;
+/// use std::panic::{UnwindSafe, RefUnwindSafe};
+/// // Rust does not allow multiple non-auto traits in trait objects, so
+/// // this is one workaround.
+/// pub trait DartDebug: DartSafe + Debug {}
+/// impl<T: DartSafe + Debug> DartDebug for T {}
+/// pub struct DebugWrapper(pub Opaque<Box<dyn DartDebug>>);
+/// // creating a DebugWrapper using the opaque_dyn macro
+/// let wrap = DebugWrapper(opaque_dyn!("foobar"));
+/// // it's possible to name it directly
+/// pub struct DebugWrapper2(pub Opaque<Box<dyn Debug + Send + Sync + UnwindSafe + RefUnwindSafe>>);
+/// ```
+#[repr(transparent)]
+#[derive(Debug)]
+pub struct Opaque<T: ?Sized + DartSafe> {
+    pub(crate) ptr: Option<Arc<T>>,
+}
+
+impl<T: ?Sized + DartSafe> From<Arc<T>> for Opaque<T> {
+    fn from(ptr: Arc<T>) -> Self {
+        Self { ptr: Some(ptr) }
+    }
+}
+
+impl<T: DartSafe> Opaque<T> {
+    pub fn new(value: T) -> Self {
+        Self {
+            ptr: Some(Arc::new(value)),
+        }
+    }
+}
+
+impl<T: DartSafe> Opaque<sync::RwLock<T>> {
+    #[inline]
+    pub fn read(&self) -> Option<sync::LockResult<sync::RwLockReadGuard<T>>> {
+        self.as_deref().map(sync::RwLock::read)
+    }
+    #[inline]
+    pub fn write(&self) -> Option<sync::LockResult<sync::RwLockWriteGuard<T>>> {
+        self.as_deref().map(sync::RwLock::write)
+    }
+    #[inline]
+    pub fn try_read(&self) -> Option<sync::TryLockResult<sync::RwLockReadGuard<T>>> {
+        self.as_deref().map(sync::RwLock::try_read)
+    }
+    #[inline]
+    pub fn try_write(&self) -> Option<sync::TryLockResult<sync::RwLockWriteGuard<T>>> {
+        self.as_deref().map(sync::RwLock::try_write)
+    }
+}
+
+impl<T: DartSafe> Opaque<sync::Mutex<T>> {
+    #[inline]
+    pub fn lock(&self) -> Option<sync::LockResult<sync::MutexGuard<T>>> {
+        self.as_deref().map(sync::Mutex::lock)
+    }
+    #[inline]
+    pub fn try_lock(&self) -> Option<sync::TryLockResult<sync::MutexGuard<T>>> {
+        self.as_deref().map(sync::Mutex::try_lock)
+    }
+}
+
+impl<T: DartSafe> Opaque<T> {
+    /// Acquire a reference to the inner value, if the pointer has not already
+    /// been disposed by Dart.
+    pub fn as_deref(&self) -> Option<&T> {
+        self.ptr.as_deref()
+    }
+}
+
+/// Dropper opaque data.
+///
+/// # Safety
+///
+/// This function should never be called manually.
+extern "C" fn drop_arc<T>(ptr: *const T) {
+    // Dart has ownership of this copy of Arc,
+    // and can only lend out clones, so this is safe to call
+    // exactly once.
+    unsafe {
+        Arc::decrement_strong_count(ptr);
+    }
+}
+
+/// Equivalent to a [`Arc::clone()`], but direcly in terms of raw pointers.
+///
+/// # Safety
+///
+/// This function should never be called manually.
+extern "C" fn lend_arc<T>(ptr: *const T) -> *const T {
```

Comment:
what if we name it `share`?

Reply:
Not sure. I personally prefer `clone` since it is indeed cloning the underlying pointer (i.e. incr ref count) and follows rust clone semantics, but if you like `share` looks also ok